src/shared/transform.rs

```diff
@@ -1,70 +1,330 @@
+        if n.kind == VarDeclKind::Const{
+            for decl in &n.decls {
+                self.collect_pat(&decl.name, decl.init.clone().map(|v| *v));
+            }
+        }
+        n.visit_children_with(self);
+    }
+
+    fn visit_fn_decl(&mut self, f: &FnDecl) {
+        self.function_bindings.insert(f.ident.to_id());
+    }
+
+}
+pub struct ThisBlockVisitor {
+    uid_identifier_map: HashMap<String, usize>
+}
+
+impl ThisBlockVisitor {
+    pub fn new() -> Self {
+        Self { uid_identifier_map: HashMap::new() }
+    }
+
+    pub fn generate_uid_identifier(&mut self, name: &str) -> Ident {
+        let name = if name.starts_with("_") {
+            name.to_string()
+        } else {
+            "_".to_string() + name
+        };
+        if let Some(count) = self.uid_identifier_map.get_mut(&name) {
+            *count = *count + 1;
+            return private_ident!(format!("{name}{count}"));
+        } else {
+            self.uid_identifier_map.insert(name.clone(), 1);
+            return private_ident!(name);
+        }
+    }
+}
+
+impl VisitMut for ThisBlockVisitor {
+
+    fn visit_mut_block_stmt(&mut self, block: &mut BlockStmt) {
+        let mut jsx_visitor = JSXVisitor {has_jsx: false };
+        block.visit_children_with(&mut jsx_visitor);
+        if jsx_visitor.has_jsx {
+            let mut this_block_visitor = ThisVisitor {
+                this_id: None,
+                this_block_visitor: self
+            };
+            block.visit_mut_children_with(&mut this_block_visitor);
+            if let Some(id) = this_block_visitor.this_id {
+                block.stmts.insert(0, Stmt::Decl(Decl::Var(Box::new(VarDecl { 
+                    span: DUMMY_SP, 
+                    kind: VarDeclKind::Const, 
+                    declare: false, 
+                    decls: vec![VarDeclarator { 
+                        span: DUMMY_SP, 
+                        name: Pat::Ident(id.into()), 
+                        init: Some(Box::new(Expr::This(ThisExpr { span: DUMMY_SP }))), 
+                        definite: false
+                    }] 
+                }))))
+            }
+        }
+    }
+
+}
+
+struct JSXVisitor {
+    has_jsx: bool
+}
+impl Visit for JSXVisitor {
+    fn visit_jsx_element(&mut self, _: &JSXElement) {
+        self.has_jsx = true;
+    }
+    fn visit_jsx_fragment(&mut self,_: &JSXFragment) {
+        self.has_jsx = true;
+    }
+}
+
+struct ThisVisitor<'a> {
+    this_id: Option<Ident>,
+    this_block_visitor: &'a mut ThisBlockVisitor
+}
+
+impl VisitMut for ThisVisitor<'_> {
+    fn visit_mut_expr(&mut self,n: &mut Expr) {
+        if let Expr::This(_) = n {
+            if self.this_id.is_none() {
+                self.this_id = Some(self.this_block_visitor.generate_uid_identifier("self$"));
+            }
+            *n = Expr::Ident(self.this_id.clone().unwrap());
+        } else {
+            n.visit_mut_children_with(self);
+        }
+    }
+}
```

Comment:
```suggestion
pub struct ThisBlockVisitor {
    this_id: Option<Ident>,
    has_jsx: bool,
}

impl ThisBlockVisitor {
    pub fn new() -> Self {
        Self {
            this_id: None,
            has_jsx: false,
        }
    }
}

impl VisitMut for ThisBlockVisitor {
    fn visit_mut_block_stmt(&mut self, block: &mut BlockStmt) {
        let mut id = self.this_id.clone();
        self.has_jsx = false;
        block.visit_mut_children_with(self);
        std::mem::swap(&mut self.this_id, &mut id);
        if let Some(id) = id {
            block.stmts.insert(
                0,
                Stmt::Decl(Decl::Var(Box::new(VarDecl {
                    span: DUMMY_SP,
                    kind: VarDeclKind::Const,
                    declare: false,
                    decls: vec![VarDeclarator {
                        span: DUMMY_SP,
                        name: Pat::Ident(id.into()),
                        init: Some(Box::new(Expr::This(ThisExpr { span: DUMMY_SP }))),
                        definite: false,
                    }],
                }))),
            )
        }
    }
    fn visit_mut_jsx_element(&mut self, el: &mut JSXElement) {
        self.has_jsx = true;
        el.visit_mut_children_with(self);
    }
    fn visit_mut_jsx_fragment(&mut self, el: &mut JSXFragment) {
        self.has_jsx = true;
        el.visit_mut_children_with(self);
    }
    fn visit_mut_expr(&mut self, n: &mut Expr) {
        if let Expr::This(_) = n {
            if !self.has_jsx {
                return;
            }
            if self.this_id.is_none() {
                self.this_id = Some(private_ident!("_self$"));
            }
            *n = Expr::Ident(self.this_id.clone().unwrap());
        } else {
            n.visit_mut_children_with(self);
        }
    }
}
```

Reply:
Thank you for fixing this issue. Nice code.